SwiftCompilerSources/Sources/Optimizer/ModulePasses/ReadOnlyGlobalVariables.swift

```diff
@@ -0,0 +1,75 @@
+//===--- ReadOnlyGlobalVariables.swift ------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SIL
+
+/// Marks global `var` variables as `let` if they are never written.
+///
+/// Note that this pass relies on the initialize-static-globals pass which converts lazily
+/// initialized globals to statically initialized globals.
+/// This pass does not mark lazily initialized globals as `let`, because such globals _are_
+/// written: in their initializers.
+///
+let readOnlyGlobalVariablesPass = ModulePass(name: "read-only-global-variables") {
+  (moduleContext: ModulePassContext) in
+
+  var writtenGlobals = Set<GlobalVariable>()
+
+  for f in moduleContext.functions {
+    for inst in f.instructions {
+      if let gAddr = inst as? GlobalAddrInst {
+        if gAddr.addressHasWrites {
```

Comment:
The most important part of this algorithm is that it performs a def-use traversal. Hiding the most important fact that code needs to convey is the definition of unreadable code.